### PR TITLE
add org.libvips.vipsdisp

### PR DESCRIPTION
**Pull request by jcupitt**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: manifest.json [hunk 1]
```diff
@@ -0,0 +1,360 @@
+{
+  "id" : "org.libvips.vipsdisp",
+  "runtime" : "org.gnome.Sdk",
+  "runtime-version" : "40",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:4]: Ah I didn't know there was a new one, changed, thank you.

File: manifest.json [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+  "sdk-commit" : "10c8aa56b67c4db4f2b1c1cec263903f786949ec1619e082681deec1b391616d",
+  "command" : "vipsdisp",
+  "modules" : [
+    {
+      "name" : "libtiff",
+      "buildsystem" : "autotools",
+      "builddir" : true,
+      "sources" : [
+        {
+          "url" : "https://download.osgeo.org/libtiff/tiff-4.3.0.tar.gz",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:16]: No, it's been built without support for JPEG compression, unfortunately.

File: manifest.json [hunk 3]
```diff
@@ -0,0 +1,360 @@
+{
+  "id" : "org.libvips.vipsdisp",
```
Comment (hfiguiere) [inline manifest.json:2]: and the manifest file should be named afer the app-id. I didn't notice that's why the build failed.

File: manifest.json [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+  "runtime" : "org.gnome.Sdk",
+  "runtime-version" : "40",
+  "runtime-commit" : "10c8aa56b67c4db4f2b1c1cec263903f786949ec1619e082681deec1b391616d",
+  "sdk" : "org.gnome.Sdk",
+  "sdk-commit" : "10c8aa56b67c4db4f2b1c1cec263903f786949ec1619e082681deec1b391616d",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:7]: I don't know where those lines came from -- I didn't make them. Perhaps the change to SDK4.1 will fix this?

File: manifest.json [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        {
+          "url" : "https://download.osgeo.org/libtiff/tiff-4.3.0.tar.gz",
+          "sha256" : "0e46e5acb087ce7d1ac53cf4f56a09b221537fc86dfc5daaad1c2e89e1b37ac8",
+          "type" : "archive"
+        }
+      ]
+    },
+    {
+      "name" : "libjxl",
+      "buildsystem" : "cmake",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:24]: Oh cool, I didn't know about that.

File: manifest.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (hfiguiere) [reply manifest.json:16]: the GNOME SDK 41 uses freedesktop-sdk 21.08 which has this fixed:

https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1117

File: manifest.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (hfiguiere) [reply manifest.json:7]:
```suggestion
  "sdk" : "org.gnome.Sdk",
```

File: manifest.json [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+      "sources" : [
+        {
+          "url" : "https://github.com/AcademySoftwareFoundation/openexr.git",
+          "commit" : "81e3d1beffb27448f6d610ed1b1cc1c962d51011",
+          "type" : "git"
+        }
+      ]
+    },
+    {
+      "name" : "pkg-config",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:131]: Yes, it's very unfortunate. 

I want openslide, but unfortunately the package lost it's maintainer in 2015 (!!):

https://openslide.org/

Due to the world moving on, it has become rather a hard package to build. You need to patch configure.ac and rerun autotools, and that means you also need a version of autotools from back then. And that means we have to build pkg-config to get all the autotools macros.

[the remark this answers is not in the document]
Comment (hfiguiere) [reply manifest.json:131]: and all of that is in the SDK

File: manifest.json [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        "-DENABLE_CMS='lcms2'",
+        "-DENABLE_LIBOPENJPEG='openjpeg2'",
+        "-DENABLE_DCTDECODER='libjpeg'",
+        "-DENABLE_SPLASH=OFF",
+        "-DENABLE_BOOST=OFF",
+        "-DENABLE_CPP=OFF"
+      ],
+      "sources" : [
+        {
+          "url" : "https://anongit.freedesktop.org/git/poppler/poppler.git",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:100]: I'd completely forgotten about poppler-data, I'll add it and swap to tarballs, thanks.

File: manifest.json [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+    "/bin/vipsthumbnail",
+    "*.la",
+    "*.a"
+  ],
+  "finish-args" : [
+    "--device=dri",
+    "--share=ipc",
+    "--socket=fallback-x11",
+    "--socket=wayland",
+    "--talk-name=org.gtk.vfs",
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:355]: OK, changed.

File: manifest.json (hunk 3 above)
Comment (jcupitt) [reply manifest.json:2]: Ah I copied `build-dir/files/manifest.json` for this PR, I was confused.

I'll copy my `org.libvips.vipsdisp.json` instead.

File: manifest.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json:131]: But not autoconf 2.69 / automake 1.14, which is the combo that openslide needs (as far as I know).

[the remark this answers is not in the document]
Comment (hfiguiere) [reply manifest.json:131]: autoconf 2.69 is in the GNOME 40 runtime you were using. (2.71 in 41)

but there are more recent than 2015 so I don't see why it would be required by this.

we build pacakge much older than that.

File: manifest.json [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+          "sha256" : "fed08fab8a9b1ded95a34e196652291127ebe392c11f9bc13d26e760295a102d",
+          "type" : "archive"
+        },
+        {
+          "path" : "patches/openslide-3-fixes.patch",
+          "type" : "patch"
+        },
+        {
+          "path" : "patches/openslide-remove-configure.patch",
+          "type" : "patch"
```
Comment (hfiguiere) [inline manifest.json:185]: these patches are missing.

File: manifest.json [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+      "builddir" : true,
+      "sources" : [
+        {
+          "url" : "https://github.com/strukturag/libde265/releases/download/v1.0.8/libde265-1.0.8.tar.gz",
+          "sha256" : "24c791dd334fa521762320ff54f0febfd3c09fc978880a8c5fbc40a88f21d905",
+          "type" : "archive"
+        },
+        {
+          "path" : "patches/libde265-1-fixes.patch",
+          "type" : "patch"
```
Comment (hfiguiere) [inline manifest.json:249]: this patch is missing too

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: manifest.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json]: This should be resolved now I've uploaded the correct manifest file.

File: manifest.json (hunk 9 above)
Comment (jcupitt) [reply manifest.json]: OK, all patches added.

File: manifest.json (hunk 10 above)
Comment (jcupitt) [reply manifest.json]: Added now.

File: org.libvips.vipsdisp.json [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+                    "url" : "https://github.com/openslide/openslide/releases/download/v3.4.1/openslide-3.4.1.tar.gz",
+                    "sha256" : "fed08fab8a9b1ded95a34e196652291127ebe392c11f9bc13d26e760295a102d"
+                },
+                {
+                    "type" : "patch",
+                    "path" : "patches/openslide-3-fixes.patch"
+                },
+                {
+                    "type" : "patch",
+                    "path" : "patches/openslide-remove-configure.patch"
```
Comment (hfiguiere) [inline org.libvips.vipsdisp.json:264]: this patch is not needed. There is the `rm-configure` option to remove the configure file.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: manifest.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (jcupitt) [reply manifest.json]: You're right, I removed auto* and pkg-config and it's building now. I'm sure it didn't work --- perhaps SDK41 changed something? Anyway, I've removed it now, thank you.

File: org.libvips.vipsdisp.json (hunk 11 above)
Comment (jcupitt) [reply org.libvips.vipsdisp.json:222]: Oh neat, I missed that. Updated.

File: org.libvips.vipsdisp.json [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+            "buildsystem" : "cmake-ninja",
+            "sources" : [
+                {
+                    "type" : "archive",
+                    "url" : "https://heasarc.gsfc.nasa.gov/FTP/software/fitsio/c/cfitsio-3.49.tar.gz",
+                    "sha256" : "5b65a20d5c53494ec8f638267fca4a629836b7ac8dd0ef0266834eab270ed4b3"
+                },
+                {
+                    "type" : "patch",
+                    "path" : "patches/cfitsio-3-fixes.patch"
```
[the remark this answers is not in the document]
Comment (jcupitt) [reply org.libvips.vipsdisp.json:273]: Thanks, added.

**Comment by jcupitt** [comment]
bot, build org.libvips.vipsdisp

File: org.libvips.vipsdisp.json [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+                    "tag" : "v2.4.0",
+                    "commit" : "37ac30ceff6640bbab502388c5e0fa0bff23f505"
+                }
+            ]
+        },
+
+        {
+            "name" : "libimagequant",
+            "builddir" : true,
+            "buildsystem" : "cmake-ninja",
```
Comment (hfiguiere) [inline org.libvips.vipsdisp.json:99]:
```suggestion
            "buildsystem" : "cmake-ninja",
            "build-options": {
                "arch": {
                    "aarch64": {
                        "config-opts": [
                            "-DBUILD_WITH_SSE=NO"
                        ]
                    }
                }
            }
```

tl;dr, another library that assume they build for an x86_64 platform.

Comment (hfiguiere) [reply org.libvips.vipsdisp.json:99]: this is untested.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
the build failed on x86_64 because it can't find the icon

File: org.libvips.vipsdisp.json (hunk 13 above)
Comment (jcupitt) [reply org.libvips.vipsdisp.json:99]: That's great! I'll try it.

**Comment by jcupitt** [comment]
I have a 512x512 icon and it wants 128x128. I'll change it.

**Comment by jcupitt** [comment]
bot, build org.libvips.vipsdisp

**Comment by jcupitt** [comment]
bot, build org.libvips.vipsdisp

**Comment by jcupitt** [comment]
It still can't find the icon :( I'm baffled. 

My Makefile.am is installing it like this:

```
vipsdispdir = $(datadir)/icons/hicolor/128x128/apps
vipsdisp_DATA = org.libvips.vipsdisp.png
```

It appears in my build-dir:

```
$ flatpak-builder --force-clean --user --install build-dir org.libvips.vipsdisp.json
...
$ ls build-dir/files/share/icons/hicolor/128x128/apps/
org.libvips.vipsdisp.png
```

And if I start the application locally, the icon is in my About dialog:

```
$ flatpak run org.libvips.vipsdisp ~/pics/k2.jpg
... select About, icon is in dialog.
```

Is there something else I need to do? I've tried reading some other manifests and a few of them use post-install to copy a PNG. Do I need to add something like that?

File: org.libvips.vipsdisp.json [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+            "buildsystem" : "autotools",
+            "build-options" : {
+                "env" : {
+                    "V" : "0"
+                }
+            },
+            "sources" : [
+                {
+                    "type" : "git",
+                    "url" : "https://github.com/jcupitt/vipsdisp.git"
```
Comment (hfiguiere) [inline org.libvips.vipsdisp.json:343]: this will require a revision / tag.

File: org.libvips.vipsdisp.json [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+        "--talk-name=org.gtk.vfs.*",
+        "--filesystem=xdg-run/gvfsd",
+        "--filesystem=home"
+    ],
+    "cleanup" : [
+        "/include",
+        "/lib/girepository-1.0",
+        "/lib/pkgconfig",
+        "/lib/cmake",
+        "/lib/debug",
```
Comment (hfiguiere) [inline org.libvips.vipsdisp.json:21]: debug should be kept. It's separated into its own flatpak `.Debug`

File: org.libvips.vipsdisp.json [hunk 16]
```diff
@@ -0,0 +1,349 @@
+{
+    "app-id" : "org.libvips.vipsdisp",
+    "runtime" : "org.gnome.Sdk",
+    "runtime-version" : "41",
+    "sdk" : "org.gnome.Sdk",
+    "command" : "vipsdisp",
```
Comment (hfiguiere) [inline org.libvips.vipsdisp.json:6]:
```suggestion
    "command" : "vipsdisp",
    "rename-desktop-file": "vipsdisp.desktop",
```

The desktop file isn't properly named. So this will take care of it.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: org.libvips.vipsdisp.json (hunk 16 above)
Comment (jcupitt) [reply org.libvips.vipsdisp.json:6]: Ahhhh, brilliant, thank you. I'll try that.

File: org.libvips.vipsdisp.json (hunk 14 above)
Comment (jcupitt) [reply org.libvips.vipsdisp.json:343]: ok, done

**Comment by jcupitt** [comment]
bot, build org.libvips.vipsdisp

**Comment by barthalion** [comment]
bot, build org.libvips.vipsdisp

**Comment by barthalion** [comment]
OK, looks like the icon file is actually there and it's not related to how the desktop file is named. Few remarks, though:

The desktop file currently looks lie this:

```
[Desktop Entry]
Type=Application
Name=Vipsdisp
Icon=org.libvips.vipsdisp
StartupNotify=true
Exec="\\${exec_prefix}/bin/vipsdisp"
```

I assume the exec_prefix isn't substituted correctly? Either way, it's safe to change it to `vipsdisp`, it doesn't have to be an absolute path.

The icon is not being exported correctly, because the appdata file has this line:

```
  <launchable type="desktop-id">vipsdisp.desktop</launchable>
```

It should be changed to `org.libvips.vipsdisp.desktop`.

File: org.libvips.vipsdisp.json [hunk 17]
```diff
@@ -0,0 +1,351 @@
+{
+    "app-id" : "org.libvips.vipsdisp",
```
Comment (barthalion) [inline org.libvips.vipsdisp.json:2]: Is it "official" libvips.org project, i.e. will be eventually moved under libvips organization on GitHub? If not, it would be better to use com.github.jcupitt.vipsdisp here (it also implies renaming icons/desktop/appdata files).

File: org.libvips.vipsdisp.json [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+    "command" : "vipsdisp",
+    "rename-desktop-file": "vipsdisp.desktop",
+    "finish-args" : [
+        "--device=dri",
+        "--share=ipc",
+        "--socket=fallback-x11",
+        "--socket=wayland",
+        "--talk-name=org.gtk.vfs.*",
+        "--filesystem=xdg-run/gvfsd",
+        "--filesystem=home"
```
Comment (barthalion) [inline org.libvips.vipsdisp.json:14]: Generally speaking, if you switch your code to use GtkNativeFileChooser, it will automatically use file portal and you'll be able to drop these permissions.

File: org.libvips.vipsdisp.json (hunk 17 above)
Comment (jcupitt) [reply org.libvips.vipsdisp.json:2]: I plan to move it into the libvips org once it's finally on 1.0, so I'd like to stick with the name.

File: org.libvips.vipsdisp.json (hunk 18 above)
Comment (jcupitt) [reply org.libvips.vipsdisp.json:14]: This is supposed to turn into the image view component for the nip2 rewrite for gtk4, so I'd like to be able to customize the file chooser.

**Comment by jcupitt** [comment]
Ahhhh good point, thanks. I'll fix the .desktop file and try again.

**Comment by jcupitt** [comment]
bot, build org.libvips.vipsdisp

**Comment by jcupitt** [comment]
It works! I tested the flathubbot build with a range of files (JXL, PDF, SVG, HEIC, FITS, openslide) and they all worked. The icon seems to be there. It can load and save settings, and you can save images in all the various formats too.

Thank you very much for fixing all this stuff, @barthalion and @hfiguiere, you've been great.

**Comment by barthalion** [comment]
bot, build org.libvips.vipsdisp

**Comment by barthalion** [comment]
bot, build org.libvips.vipsdisp
